### PR TITLE
Fix/xr2

### PR DESCRIPTION
**Pull request by kaychang-unity**
---
### Purpose of this PR
Reverted some change in matrix for XR mode, as they were incorrect.

---
### Testing status

**Manual Tests**: What did you do?
- [ X ] Opened test project + Run graphic tests locally

**Yamato**: (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

**Comment by kaychang-unity** [comment]
I reverted the change for _InvCameraViewProj: it is never equivalent to unity_MatrixInvVP because the former applies z adjustments, while the later does not.
However, original code reverted defines UNITY_MATRIX_I_VP as _InvCameraViewProj (which is wrong), so I just added a comment since there is no possible fix to it without adding C++ code.

File: com.unity.render-pipelines.universal/ShaderLibrary/Input.hlsl [hunk 1]
```diff
@@ -65,9 +65,13 @@ half4 _AdditionalLightsOcclusionProbes[MAX_VISIBLE_LIGHTS];
...
 #define UNITY_MATRIX_I_V   unity_MatrixInvV
 #define UNITY_MATRIX_P     OptimizeProjectionMatrix(glstate_matrix_projection)
-#define UNITY_MATRIX_I_P   ERROR_UNITY_MATRIX_I_P_IS_NOT_DEFINED
+#define UNITY_MATRIX_I_P   unity_CameraInvProjection // workaround: technically per-pass matrix set is different from camera matrix set
 #define UNITY_MATRIX_VP    unity_MatrixVP
-#define UNITY_MATRIX_I_VP  unity_MatrixInvVP
+#if defined(USING_STEREO_MATRICES)
+#define UNITY_MATRIX_I_VP  unity_MatrixInvVP // only defined in stereo-mode
+#else
+#define UNITY_MATRIX_I_VP  _InvCameraViewProj // incorrect: _InvCameraViewProj applies platform-specific z adjustments (reversed-z and [0,1] range), while unity_MatrixInvVP does not.
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/ShaderLibrary/Input.hlsl]: The apply of platform-specific adjustments should not matter because of name of the var. So I don't understand why we can't bind as unity_MatrixInvVP.

From what I look it seems to be like this. 
If we call CommandBuffer.SetViewProjectionMatrix(view, proj) -> Unity will apply platform-specific z adjustments in c++ here.

If we call SetGlobalMatrix() it won't. However, the API GL.GetGPUMatrix(proj) will apply these.
So from what I look from the code, maybe we should remove the GL.GetGPUMatrix from here as the we might been applying twice?

https://github.com/Unity-Technologies/ScriptableRenderPipeline/blob/master/com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs#L582-L586

File: com.unity.render-pipelines.universal/CHANGELOG.md [hunk 2]
```diff
@@ -61,7 +61,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Fixed issue where using a ForwardRendererData preset would cause a crash. [case 1201052](https://issuetracker.unity3d.com/product/unity/issues/guid/1201052/)
 - Fixed an issue where particles had dark outlines when blended together [case 1199812](https://issuetracker.unity3d.com/issues/urp-soft-particles-create-dark-blending-artefacts-when-intersecting-with-scene-geometry)
 - Fixed an issue with deleting shader passes in the custom renderer features list [case 1201664](https://issuetracker.unity3d.com/issues/urp-remove-button-is-not-activated-in-shader-passes-list-after-creating-objects-from-renderer-features-in-urpassets-renderer)
-- Fixed camera inverse view-projection matrix in XR mode, depth-copy and color-copy passes.
+- Fixed depth-copy and color-copy passes.
+- Fixed ERROR_UNITY_MATRIX_I_P_IS_NOT_DEFINED shader error.
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/CHANGELOG.md]: This changelog should be in shader graph as well.

File: com.unity.render-pipelines.universal/ShaderLibrary/Input.hlsl (hunk 1 above)
Comment (kaychang-unity) [reply com.unity.render-pipelines.universal/ShaderLibrary/Input.hlsl]: I don't know who added the call to GL.GetGPUMatrix(proj) but was under the impression removing it would be "bad" for existing user projects that may rely on this parameter. Anyhow, I can definitely remove it. In that case _inCameraViewProj will be equal to unity_MatrixInvVP indeed.

Comment (kaychang-unity) [reply com.unity.render-pipelines.universal/ShaderLibrary/Input.hlsl]: Fixed.

File: com.unity.render-pipelines.universal/CHANGELOG.md (hunk 2 above)
Comment (kaychang-unity) [reply com.unity.render-pipelines.universal/CHANGELOG.md]: Fixed.

**Comment by kaychang-unity** [comment]
@phi-lira Next bug I found is in XR Multi-pass.
_InvCameraViewProj/unity_MatrixInvVP is updated from C# code at
https://github.com/Unity-Technologies/ScriptableRenderPipeline/blob/621e0f29bab9435f66f4f42a07916ccd4ddcd14f/com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs#L568 
However, this is outside of the loop for rendering each eyes in multi-pass:
https://github.com/Unity-Technologies/ScriptableRenderPipeline/blob/621e0f29bab9435f66f4f42a07916ccd4ddcd14f/com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs#L234

As a result, the right eye is not correctly rendered. This issue does not affect "Single pass" and "Single Pass Instanced" because in these modes, unity_MatrixInvVP is setup from C++ code.

To sum up, the bug is that using UNITY_MATRIX_I_VP in XR Multi-pass will generate incorrect result for the right eye.

**Comment by kaychang-unity** [comment]
@phi-lira  I committed a fix to have correct UNITY_MATRIX_I_VP in XR Multi-pass with revision 
41f491d. Please advise if you want this fix or have it reverted.

**Comment by ellioman** [comment]
@kaychang-unity Can you fix the conflicts?

@phi-lira Can you re-review this and check the two comments above this one?

**Comment by kaychang-unity** [comment]
I fixed the conflicts and it compiles again.
I'll test XR multi-pass with the MockHMD tomorrow, once I figure out how to enable it again (the old UI has disappeared, now asking XR channel).

**Comment by kaychang-unity** [comment]
Multipass cannot be tested at the moment. UniversalGfxTestStereo regressed in master branch to the point only SinglePass is working (SinglePass instanced and Multipass are broken). I was told UniversalGfxTestStereo will be soon updated from using LegacyXR to the new XRSDK. 
UniversalGraphicsTests is also currently broken for XR support (using new XR SDK).

File: com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs [hunk 3]
```diff
@@ -766,24 +765,31 @@ static void SetupPerFrameShaderConstants()
             Shader.SetGlobalVector(PerFrameBuffer._SubtractiveShadowColor, CoreUtils.ConvertSRGBToActiveColorSpace(RenderSettings.subtractiveShadowColor));
         }
 
-        static void SetupPerCameraShaderConstants(in CameraData cameraData)
+        internal static void SetupPerCameraShaderConstants(ScriptableRenderContext context, CameraData cameraData, bool stereoEnabled, int eyeIndex)
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs]: Check if the ScripableRenderer already creates a command buffer we could reuse. 
Also, we might as well move this to ScriptableRenderer.

**Comment by phi-lira** [comment]
PR seems good to me, I'd like to get 
@thomas-zeng opinion/approval before merging. 

Thomas is also changing how to setup camera matrix logic here: https://github.com/Unity-Technologies/ScriptableRenderPipeline/pull/4857

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs [hunk 4]
```diff
@@ -289,6 +289,7 @@ public void Execute(ScriptableRenderContext context, ref RenderingData rendering
             /// * Setup global time properties (_Time, _SinTime, _CosTime)
             bool stereoEnabled = renderingData.cameraData.isStereoEnabled;
             context.SetupCameraProperties(camera, stereoEnabled, eyeIndex);
+            UniversalRenderPipeline.SetupPerCameraShaderConstants(context, renderingData.cameraData, stereoEnabled, eyeIndex);
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs]: This change will introduce a regression because some of these properties are required to be set before, f.ex, _WorldSpaceCameraPos.

Maybe we need to split this into two functions. One for setting up just camera matrices, other for everything else.

Comment (kaychang-unity) [reply com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs]: I split it to 2 functions:
- original function is back where it was with most of the original code.
-  a new SetupPerCameraMatrices() that only update unity_MatrixInvVP.
Do you still want this function to be moved to ScriptableRenderer (with different name)?

Comment (phi-lira) [reply com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs]: "Do you still want this function to be moved to ScriptableRenderer" That's fine to be like it is now. It's internal so we can always change later if neeed.

File: com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs [hunk 5]
```diff
@@ -834,12 +834,18 @@ static void SetupPerCameraShaderConstants(in CameraData cameraData)
...
+        }
+
+        internal static void SetupPerCameraMatrices(CommandBuffer cmd, ScriptableRenderContext context, CameraData cameraData, bool stereoEnabled, int eyeIndex)
+        {
+            // The following camera properties must be set per eye for XR support.
+            Camera camera = cameraData.camera;
 
-            Matrix4x4 projMatrix = GL.GetGPUProjectionMatrix(camera.projectionMatrix, false);
-            Matrix4x4 viewMatrix = camera.worldToCameraMatrix;
+            Matrix4x4 projMatrix = stereoEnabled ? camera.GetStereoProjectionMatrix((Camera.StereoscopicEye)eyeIndex) : camera.projectionMatrix;
```
Comment (thomas-zeng) [inline com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs]: In SPI and MultiView pass, camera inv projection is matrix4x4Array type. Shader indexed into this array based on eye idx (instance id in SPI mode and view id in MultiView mode).

The logic here of setting stereo proj and view only works for multipass.

Comment (thomas-zeng) [reply com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs]: I think we should either be clear this is not supported in XR at all(throw a warning and communicate XR is not supported to user)
or support multipass, SPI and Multiview fully.

Let me know what you think! :)

Comment (kaychang-unity) [reply com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs]: Yes, this code only make sense if we are in multipass mode. I can add extra logic to only run this code if we are in that mode.
I don't know what are the plans as to which modes XR wants to support long term though.

**Comment by kaychang-unity** [comment]
@thomas-zeng Does the latest commit addresses your concern?

File: com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs (hunk 5 above)
Comment (thomas-zeng) [reply com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs]: URP long term will support:
Multiview, SPI, Multipass 
Android platforms that supports multiview extension eg, Quest, MagicLeap run Multiview
PC&Console eg WMR, rift run SPI

Multi-pass is a fallback only.

**Comment by thomas-zeng** [comment]
I am still having concerns with the latest commit. What if the user is on SPI/Multiview which I expect most XR users will be on? Does the shader using inv camera projection work for them?
If not, we need to emit not supported message to them otherwise they will see broken rendering result and wonder what's going wrong.

**Comment by kaychang-unity** [comment]
Please confirm whether my understanding is correct. When SPI/Multiview is enabled:
- `IsMultiPassStereoEnabled(camera)` return false
- `USING_STEREO_MATRICES` is defined in shader, so the following define will be used:
`#define unity_MatrixInvVP mul(unity_StereoMatrixInvV[unity_StereoEyeIndex], unity_StereoCameraInvProjection[unity_StereoEyeIndex])`

If that's correct, then I think it should be fine.
I am not really able to test XR multipass myself at the moment (I don't have an actual XR device and new mockHMD is still not working in URP, right?).

**Comment by thomas-zeng** [comment]
Hey, thanks kay for your explanation.
With the shader ifdef USING_STEREO_MATRICES logic, I believe XR should be fine.

MockHMD should be working in URP now. 
We should also have XR QA coverage on effect that is using unity_MatrixInvVP.
Could you suggest which test case we should cover here?

**Comment by thomas-zeng** [comment]
Adding XR QA as reviewer

**Comment by kaychang-unity** [comment]
Thanks, I will try running the mockHMD again tomorrow.

For testing, I don't have a naturally occurring test case. However, I can artificially make the deferred renderer use `UNITY_MATRIX_I_VP` in XR Multipass.
If you replace the line at https://github.com/Unity-Technologies/ScriptableRenderPipeline/blob/289e6f5658587029f8bedeb74ff8585f45c70b97/com.unity.render-pipelines.universal/Shaders/Utils/StencilDeferred.shader#L103 with `#if 0`, you can force deferred shading code to take the pass that use `UNITY_MATRIX_I_VP`.

**Comment by phi-lira** [comment]
@DennisDeRykeUnity any update on this?

**Comment by DennisDeRykeUnity** [comment]
I see that "_ABV for SRP repository - fast-trunk" is green on Yamato (branch: fix/xr2, commit: 3176f753): 
https://yamato.prd.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline/tree/fix%252Fxr2/.yamato%252Fupm-ci-abv.yml%2523all_project_ci_fast-trunk/1325921/job

I tested this build using: 
- XRSDK
- URP
- URP Template scene, with exaggerated Bloom Tint and Intensity
- Oculus Rift in Play mode
- MultiPass and Stereo Instancing
- Unity 2020.1.0a23

I intend to do a similar test on Oculus Go prior to approving this PR.  

@kaychang-unity Beyond this, is there any additional XR testing that you recommend?  Is there some specific shader that I should use to exercise your code changes?

**Comment by kaychang-unity** [comment]
@DennisDeRykeUnity Not particularly. But unrelated to XR support, this should fix this fogbugz case: https://fogbugz.unity3d.com/f/cases/1115116

**Comment by phi-lira** [comment]
Missing shader graph reviewers.

File: com.unity.render-pipelines.universal/CHANGELOG.md [hunk 6]
```diff
@@ -76,6 +76,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Fixed an issue where particles had dark outlines when blended together [case 1199812](https://issuetracker.unity3d.com/issues/urp-soft-particles-create-dark-blending-artefacts-when-intersecting-with-scene-geometry)
 - Fixed an issue with deleting shader passes in the custom renderer features list [case 1201664](https://issuetracker.unity3d.com/issues/urp-remove-button-is-not-activated-in-shader-passes-list-after-creating-objects-from-renderer-features-in-urpassets-renderer)
 - Fixed camera inverse view-projection matrix in XR mode, depth-copy and color-copy passes.
+- Fixed _InvCameraViewProj shader parameter to not be calculated from GL.GetGPUMatrix(proj) anymore.
```
Comment (pbbastian) [inline com.unity.render-pipelines.universal/CHANGELOG.md]: Unmatched underscore messes up the markdown formatting for the rest of the file.
```suggestion
- Fixed `_InvCameraViewProj` shader parameter to not be calculated from GL.GetGPUMatrix(proj) anymore.
```

File: com.unity.shadergraph/CHANGELOG.md [hunk 7]
```diff
@@ -15,6 +15,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Updated the zoom level to let you zoom in further.
 - If Unity Editor Analytics are enabled, Shader Graph collects anonymous data about which nodes you use in your graphs. This helps the Shader Graph team focus our efforts on the most common graph scenarios, and better understand the needs of our customers. We don't track edge data and cannot recreate your graphs in any form.
 - The Create Node Menu now has a tree view and support for fuzzy field searching.
+- Fixed a `ERROR_UNITY_MATRIX_I_P_IS_NOT_DEFINED` shader error.
```
Comment (pbbastian) [inline com.unity.shadergraph/CHANGELOG.md]: Since this doesn't have any changes to Shader Graph itself, the change log item should go into the Universal change log.

Comment (kaychang-unity) [reply com.unity.shadergraph/CHANGELOG.md]: I reverted that line of code.

**Comment by kaychang-unity** [comment]
@LandonTownsendUnity I confirm that  UNITY_MATRIX_I_VP **IS NOT** the inverse of UNITY_MATRIX_VP. As to how it came out to be this way .... I dont know.
I think UNITY_MATRIX_I_VP  does not include platform specific adjustment (UNITY_REVERSED_Z, UNITY_UV_STARTS_AT_TOP and NDC z half/full unit cube), while UNITY_MATRIX_VP does, which is why these multiplying these matrices together does not give identity.
@phi-lira we have similar issue with UNITY_MATRIX_P and UNITY_MATRIX_I_P, where if we define 
UNITY_MATRIX_I_P == unity_CameraInvProjection, then UNITY_MATRIX_I_P **IS NOT** the inverse of UNITY_MATRIX_P.

**Comment by phi-lira** [comment]
@kaychang-unity while looking at y-flip issues in URP I've realized it's an oddly hard situation to fix if we don't handle camera matrices fully in URP side (due to engine SetRenderTarget, SetViewProjectionMatrices, Blit and SetCameraProperties having logic to patch projection matrix.)
So I started working on it #5913

I could take this PR and merge into #5913 or we wait until #5913 lands to land this one. What do you think it's best?

**Comment by kaychang-unity** [comment]
@phi-lira I would prefer if you take this PR and merge it into #5913 . As you said we need URP to fully handle the camera matrices to solve the issue properly. It also means that #5913 will also have the fix for https://fogbugz.unity3d.com/f/cases/1115116 .
